components/external_storage/src/s3.rs

```diff
@@ -387,6 +440,25 @@ mod tests {
         let ret = block_on_external_io(reader.read_to_end(&mut buf));
         assert!(ret.unwrap() == 0);
         assert!(buf.is_empty());
+
+        // test timeout
+        let s3_timeout_injected_fp = "s3_timeout_injected";
+        let s3_sleep_injected_fp = "s3_sleep_injected";
+
+        // inject 1 second timeout
+        fail::cfg(s3_timeout_injected_fp, "return(1)").unwrap();
+        // inject 2 seconds delay
+        fail::cfg(s3_sleep_injected_fp, "return(2)").unwrap();
+
+        let resp = s.write(
+            "mykey2",
+            Box::new(magic_contents.as_bytes()),
+            magic_contents.len() as u64,
+        );
+
```

Comment:
You need to check if the duration is smaller than the delay time.
And one second is a little long for a test, maybe 100ms is enough